finality-aleph/src/metrics/slo.rs

```diff
@@ -0,0 +1,155 @@
+use std::fmt::Debug;
+
+use derive_more::Display;
+use futures::{Stream, StreamExt};
+use log::{info, warn};
+use parity_scale_codec::Encode;
+use primitives::Block;
+use sc_client_api::{FinalityNotifications, ImportNotifications};
+use sp_consensus::BlockOrigin;
+use sp_runtime::traits::Block as _;
+use substrate_prometheus_endpoint::Registry;
+
+use super::{finality_rate::FinalityRateMetrics, timing::DefaultClock};
+use crate::{
+    block::ChainStatus,
+    metrics::{
+        best_block_related::BestBlockRelatedMetrics, timing::Checkpoint,
+        transaction_pool::TransactionPoolMetrics, TimingBlockMetrics, LOG_TARGET,
+    },
+    BlockId, SubstrateChainStatus,
+};
+
+#[derive(Debug, Display)]
+pub enum Error {
+    BlockImport,
+    FinalityNotification,
+    TransactionPool,
+}
+
+pub async fn run_metrics_service<TS: Stream<Item = TxHash> + Unpin>(
+    metrics: &SloMetrics,
+    import_notifications: &mut ImportNotifications<Block>,
+    finality_notifications: &mut FinalityNotifications<Block>,
+    transaction_pool_stream: &mut TS,
+) -> Result<(), Error> {
+    if metrics.is_noop() {
+        info!(target: LOG_TARGET, "Stopping metrics service: metrics were disabled.");
+        return Ok(());
+    }
+    loop {
+        tokio::select! {
+            maybe_block = import_notifications.next() => {
```

Comment:
If you are already using import/finality notifications then imo it would be better to have this happen in `sync`. In particular the whole business with figuring out the length of the reorg should become much simpler (no need for the trait and whatnot), but also it would better reflect the nodes internal notion of best (or rather favourite) block.

The runner here would still be needed for transactions though.

Reply:
The source-of-truth for what "favourite" is is actually in the forest, so that happens on receiving an import notification by definition. The info you use now is afaik less accurate (I think it might even treat all reaorgs as reorgs to the top finalized actually, although I haven't checked).

After a quick glance into the forest I'm a tad more worried about getting the reorg length from there though, it might not be as trivial as I assumed. :/

Reply:
Ah, ok. And reorgs could be detected by looking into db for the lca when updating favorite block, right?

Reply:
Yeah, although I'm slightly worried about this call in general, DB operations like this turned out to be particularly costly in the past, so it would be nice to not have them done here. Looking for lca in the forest should be simple to implement and much faster. On the other hand while there _used_ to be problems with the DB, I doubt there will be now, since we shouldn't be getting reorgs of length >20 ever, and that _cannot_ take long, so maybe it's not worth it.

tl;dr do as you with with reorg length calculation :P

Reply:
> The info you use now is afaik less accurate (I think it might even treat all reaorgs as reorgs to the top finalized actually, although I haven't checked).

If I'm not mistaken, I think I might be missing some temporary reorgs (succeeded by another reorg shortly after), but their length should be fine. Also, info from import notification streams is what frontend sees, so it is more important than what happens locally in the forest. 
Also, a slight improvement to this would be to detect reorgs also on every favorite block request by select chain - this way we won't miss any reorg visible from outside of sync.

Reply:
With the changes that @kostekIV made to best block detection recently I think the approach you implemented (computing the length of the reorg exactly when the DB decides a newly imported block is the new best block) will slightly(?) overestimate the length of reorgs. But my intuition here is far from perfect, what do you think Jan?

I also think that changing the return values just requires writing proper types for them, and then it shouldn't be that bad? But no pressure on that point, if you didn't like what this looked like, it can stay as it is currently.

Reply:
Yep they will be overestimated. Its because the best block can change during finalization and we will compute the lenght of the reorg for imported blocks after real reorg happened

Reply:
Hmm, but the length of the reorg is the length of the retracted branch. The lowest common ancestor of the last imported block pre reorg caused by finalization, and any next imported block is the same, so the retracted part length also is, no?

Reply:
Oh, yeah, that should work, fair enough.